lttoolbox/lt_paradigm.cc

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright (C) 2022 Apertium
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, see <https://www.gnu.org/licenses/>.
+ */
+#include <lttoolbox/alphabet.h>
+#include <lttoolbox/file_utils.h>
+#include <lttoolbox/input_file.h>
+#include <lttoolbox/lt_locale.h>
+#include <lttoolbox/state.h>
+#include <lttoolbox/trans_exe.h>
+
+#include <getopt.h>
+#include <iostream>
+#include <libgen.h>
+#include <queue>
+
+void endProgram(char* name)
+{
+  std::cout << basename(name) << ": generate listings from a compiled transducer" << std::endl;
+  std::cout << "Usage: " << basename(name) << " [ -a ] FST [ input [ output ] ]" << std::endl;
+  std::cout << "  -a, --analyser:       FST is an analyser (tags on the right)" << std::endl;
+  std::cout << "  -h, --help:           Print this help and exit" << std::endl;
+  exit(EXIT_FAILURE);
+}
+
+void expand(Transducer& inter, int state, const std::set<int>& past_states,
+            const std::vector<int32_t>& syms, const Alphabet& alpha, UFILE* out)
+{
+  if (inter.isFinal(state)) {
+    UString l, r;
+    for (auto& it : syms) {
+      auto pr = alpha.decode(it);
+      alpha.getSymbol(l, pr.first);
+      alpha.getSymbol(r, pr.second);
+    }
+    if (!l.empty() && !r.empty()) {
+      u_fprintf(out, "%S:%S\n", r.c_str(), l.c_str());
+    }
+  }
+  std::set<int> new_states = past_states;
+  new_states.insert(state);
+  for (auto& it : inter.getTransitions()[state]) {
+    if (past_states.find(it.second.first) != past_states.end()) {
+      continue;
+    }
+    std::vector<int32_t> new_syms = syms;
+    new_syms.push_back(it.first);
+    expand(inter, it.second.first, new_states, new_syms, alpha, out);
+  }
+}
+
+void process(const UString& pattern, std::map<UString, Transducer>& trans,
+             Alphabet& alpha,
+             const std::set<UChar32>& letters, const std::set<int32_t> tags,
```

Comment:
`const std::set<int32_t>` is weird. Missing a `&`?

Reply:
yep, fixed